### ISSUE TITLE
Increase unit test coverage and add support for codecov.io

### ISSUE THREAD

**Issue by simsong** [sleuthkit/sleuthkit#2472]
The reliability of Sleuthkit and the ease of accepting pull requests would be improved if there was increased support for unit tests and a code coverage analysis tool, such as codecov.io. Support for this was recently added to bulk_extractor. The [bulk_extractor github Workflows ](https://github.com/simsong/bulk_extractor/tree/master/.github/workflows) show how to do this.

**Comment by bcarrier** [comment]
Hey Simson,

We were using Codacy up until recently and it stopped working since it wanted what I thought were excessive permissions. It wanted write permissions for ssh keys. 

Does codecov provide feedback on following certain "hygiene" rules or is it just coverage?

**Comment by simsong** [comment]
I don't have experience with Codacy.  My main static checker is the compiler.  I'm going to be enabling AddressSanitizer and ThreadSanitizer on bulk_extractor.

Codecov is very useful for identifying where I should expand unit test coverage. For `bulk_extractor` 2.0 I've started doing a lot of small unit tests. It's been useful. I learned this while working on some of my Python tools. At Census we did a lot of code coverage stuff, but we did it with the `pytest-cov` framework, as we didn't have Codecov on site and couldn't use the hosted version.

Codecov.io had a compromise recently. It impacted me, but not much. I use GitHub secrets to maintain both the Codecov.io token and some MySQL usernames and passwords. The hackers got the write password for a test database. I've now refactored the code such that the hackers would only get a read-only password for that test database. I do not run codecov tests with a writable database, but I do run them in a different process.

**Comment by joachimmetz** [comment]
@bcarrier codacy is basically a web UI that uses a lot of freely available tools you technically could run as part of the CI test https://docs.codacy.com/related-tools/codacy-plugin-tools/

**Comment by simsong** [comment]
I think that the way I fix the codecov problem is by having the unit test executable (newly reworked to use catch2) link the .o files directly, and not libtsk.

**Comment by uckelman-sf** [comment]
Does codecov work? The test output reprorts quite a few files it doesn't find which ought to be present. Where can I go to see the whatever the result of running codecov is?

**Comment by simsong** [comment]
It's definately doing something
https://app.codecov.io/github/sleuthkit/sleuthkit
<img width="1342" alt="image" src="https://github.com/user-attachments/assets/768439be-29c4-45e9-be32-ec37cb7d4499">

But it's not doing enough, and I'm not sure why.

**Comment by joachimmetz** [comment]
> But it's not doing enough, and I'm not sure why.

per https://app.codecov.io/github/sleuthkit/sleuthkit/commit/51d39e8394f69ed59d71e26bb7bb07f3dcad4b20

```
Unusable report due to issues such as source code unavailability, path mismatch, empty report, or incorrect data format. Please visit our [troubleshooting document](https://docs.codecov.com/docs/error-reference#unusable-reports)
 for assistance.
```

**Comment by simsong** [comment]
Yes, I've read that, but I don't understand what we're doing wrong. I suspect that it's because of some error in the YAML file, but I can't figure it out.

**Comment by simsong** [comment]
Okay, I'm going to look at https://www.codacy.com/, per @bcarrier's comments above.
